test/controllers/tradingNames/AddTradingNameControllerSpec.scala

```diff
@@ -0,0 +1,177 @@
+          .set(TradingNamePage(Index(4)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(5)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(6)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(7)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(8)), TradingName("foo")).success.value
+          .set(TradingNamePage(Index(9)), TradingName("foo")).success.value
+
+      val application = applicationBuilder(userAnswers = Some(userAnswers)).build()
+
+      running(application) {
+        val request = FakeRequest(GET, addTradingNameRoute)
+
+        val view = application.injector.instanceOf[AddTradingNameView]
+
+        val list = TradingNameSummary.addToListRows(userAnswers, waypoints, AddTradingNamePage())
+
+        val result = route(application, request).value
+
+        status(result) mustBe OK
+        contentAsString(result) mustBe view(form.fill(true), waypoints, list, canAddTradingNames = false)(request, messages(application)).toString
+      }
+    }
+
+    "must save the answer and redirect to the next page when valid data is submitted" in {
+
+      val mockSessionRepository = mock[AuthenticatedUserAnswersRepository]
+
+      when(mockSessionRepository.set(any())) thenReturn true.toFuture
+
+      val application =
+        applicationBuilder(userAnswers = Some(baseAnswers))
+          .overrides(
+            bind[AuthenticatedUserAnswersRepository].toInstance(mockSessionRepository)
+          )
+          .build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, addTradingNameRoute)
+            .withFormUrlEncodedBody(("value", "true"))
+
+        val result = route(application, request).value
+        val expectedAnswers = baseAnswers.set(AddTradingNamePage(Some(index)), true).success.value
+
+        status(result) mustBe SEE_OTHER
+        redirectLocation(result).value mustBe AddTradingNamePage(Some(index)).navigate(waypoints, baseAnswers, expectedAnswers).url
+        verify(mockSessionRepository, times(1)).set(eqTo(expectedAnswers))
+      }
+    }
+
+    "must return a Bad Request and errors when invalid data is submitted" in {
+
+      val application = applicationBuilder(userAnswers = Some(baseAnswers)).build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, addTradingNameRoute)
+            .withFormUrlEncodedBody(("value", ""))
+
+        val boundForm = form.bind(Map("value" -> ""))
+
+        val view = application.injector.instanceOf[AddTradingNameView]
+
+        val list = TradingNameSummary.addToListRows(baseAnswers, waypoints, AddTradingNamePage())
+
+        val result = route(application, request).value
+
+        status(result) mustBe BAD_REQUEST
+        contentAsString(result) mustBe view(boundForm, waypoints, list, canAddTradingNames = true)(request, messages(application)).toString
```

Comment:
If you comment out most of AddTradingNameView the test still works. 

```
@layout(pageTitle = title(form, titleText)) {

    @formHelper(action = routes.AddTradingNameController.onSubmit(waypoints), Symbol("autoComplete") -> "off") {

@*        @if(form.errors.nonEmpty) {*@
@*            @govukErrorSummary(ErrorSummaryViewModel(form))*@
@*        }*@

@*        <h1 class="govuk-heading-l">@headingText</h1>*@

@*        @addToList(tradingNames, Long, "addTradingName.change.hidden", "addTradingName.remove.hidden")*@

@*        @if(canAddTradingNames) {*@
@*            @govukRadios(*@
@*                RadiosViewModel.yesNo(*@
@*                    field = form("value"),*@
@*                    legend = LegendViewModel(*@
@*                        HtmlContent(Html("""<h2 class="govuk-fieldset__heading">""" + messages("addTradingName.addAnother") + "</h2>"))*@
@*                    ).withSize(LegendSize.Medium)*@
@*                ).withHint(HintViewModel(messages("addTradingName.addAnother.hint")))*@
@*            )*@
@*        } else {*@
@*            <p class="govuk-body">@messages("addTradingName.maximumReached")</p>*@
@*            <input type="hidden" name="value" id="value" value="false"/>*@
@*        }*@

@*        @button("site.continue", routes.AddTradingNameController.onPageLoad(waypoints).url)*@
    }
}

```
Can use.

```scala
        val doc = Jsoup.parse(contentAsString(result))
        doc.select(".govuk-fieldset__heading").text() mustBe "Add another trading name?"
```
need to play a game of find the selector. Can be nicer to just have constant vals referenced in the test and the Scala template as when things get complicated it can be pretty easy to mistype stuff. One of the positives of twirl being compilable.